core/src/main/scala/cats/data/Prod.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+/**
+ * [[Prod]] is a product to two independent functor values.
+ *
+ * See: [[https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf The Essence of the Iterator Pattern]]
+ */
+sealed trait Prod[F[_], G[_], A] {
+  def first: F[A]
+  def second: G[A]
+}
+object Prod extends ProdInstances {
+  def apply[F[_], G[_], A](first0: => F[A], second0: => G[A]): Prod[F, G, A] = new Prod[F, G, A] {
```

Comment:
Should this just take two `Lazy` args instead of forcing by-name and then converting to `Lazy` anyway?